### PR TITLE
New API

### PR DESCRIPTION
**Pull request by wojtekmach**
Major changes:

  * Add high-level functional API: `Req.new(options) |> Req.request(options)`

  * Add `Req.Request.options` field that steps can read from. Also, make
    all steps be arity 1.

    When using "High-level" API, we now run all steps by default. (The
    steps, by looking at `request.options`, can decide to be no-op.)

  * Remove `put_default_steps/2` step.

  * Remove `run_steps/2` step.

Deprecations:

  * Deprecate calling `Req.post!(url, body)` in favour of `Req.post(url,
    body: body)`. Same for `Req.put!`.

  * Deprecate setting `retry: [delay: delay, max_retries: max_retries]`
    in favour of `[retry_delay: delay, max_retries: max_retries]`.

  * Deprecate setting `follow_redirects: [location_trusted: trusted]`
    in favour of `[location_trusted: trusted]`.

  * Deprecate setting `cache: [dir: dir]` in favour of `cache_dir: dir`.

  * Deprecate `Req.Request.build/3` in favour of `Req.Request.new/1`.

### CODE REVIEW

File: lib/req.ex [hunk 1]
```diff
@@ -10,92 +10,314 @@ defmodule Req do
...
+              end
+
+            # TODO: merge headers
+
+            {%{request | option => value}, Keyword.delete(options, option)}
+
+          :error ->
+            {request, options}
+        end
+      end)
```
Comment (jonatanklosko) [inline lib/req.ex:281]:
```suggestion
      {request_options, options} = Keyword.split(options, [:method, :url, :headers, :body])

      request =
        Map.merge(request, Map.new(request_options), fn
            :url, _, url -> URI.parse(url)
            # TODO: merge headers
            :headers, _, headers -> headers
            _, _, value -> value
        end)
```

File: lib/req.ex [hunk 2]
```diff
@@ -10,92 +10,314 @@ defmodule Req do
...
+    # response steps
+    follow_redirects: true,
+    location_trusted: false,
+    raw: false,
+
+    # error steps
+    max_retries: 2,
+    retry: true,
+    retry_delay: 2000
+  }
```
Comment (jonatanklosko) [inline lib/req.ex:34]: Since we include all the steps, why can't this be empty and make each step get its options with the corresponding defaults?

**Comment by jonatanklosko** [comment]
Exciting! :cat:

File: lib/req.ex (hunk 2 above)
Comment (wojtekmach) [reply lib/req.ex:34]: Good question! I think that's actually preferable but by supplying defaults we can use guards: https://github.com/wojtekmach/req/pull/88/files#diff-07dc6fd4ec95415d610a3e3ddf9374540d6d64d02eaf42df7a233d6f05b32cc1R425. But it's probably not worth it just for guards sake, dunno. Will think about it some more.

Comment (wojtekmach) [reply lib/req.ex:34]: oops, link doesn't work, here's an example:

```elixir
  def decompress({request, response}) when request.options.raw == true do
     {request, response}
   end
```

Comment (jonatanklosko) [reply lib/req.ex:34]: If we dropped the `raw: false` default the same guard would work though, no?

Comment (wojtekmach) [reply lib/req.ex:34]: oh, guards shouldn't fail if `options` doesn't have `:raw` key. Yup, thanks!

**Pull request merged**